### PR TITLE
Google Search: Allow extracting items from SERP results

### PR DESCRIPTION
**Pull request by Gallaecio**
To do:
- [x] https://github.com/zytedata/zyte-common-items/pull/111
- [x] https://github.com/zytedata/zyte-common-items/pull/119

### CODE REVIEW

File: zyte_spider_templates/spiders/serp.py [hunk 1]
```diff
@@ -48,6 +58,83 @@ class SerpMaxPagesParam(BaseModel):
...
+class SerpItemTypeParam(BaseModel):
+    item_type: SerpItemType = Field(
+        title="Item type",
+        description="Data type of the output items.",
+        default=SerpItemType.serp,
+        json_schema_extra={
+            "enumMeta": {
+                # TODO: Add a test to make sure this is in sync with the enum class above.
+                # TODO: Try automating the generation of this metadata from the enum type above.
+                SerpItemType.serp: {
```
Comment (kmike) [inline zyte_spider_templates/spiders/serp.py:120]: Hm, that's an interesting idea. I was thinking that the default here would be None, which means "don't go to the results, output the links".

**Review by wRAR — APPROVED** [review]

File: zyte_spider_templates/spiders/serp.py [hunk 2]
```diff
@@ -48,6 +59,61 @@ class SerpMaxPagesParam(BaseModel):
...
+    SerpItemType.forumThread: ForumThread,
+    SerpItemType.jobPosting: JobPosting,
+    SerpItemType.product: Product,
+    SerpItemType.productList: ProductList,
+}
+
+
+class SerpItemTypeParam(BaseModel):
+    item_type: Optional[SerpItemType] = Field(
+        title="Item type",
```
Comment (kmike) [inline zyte_spider_templates/spiders/serp.py:107]: Hm, here, in the title, "item type" doesn't read very well, as you noticed :(

File: zyte_spider_templates/spiders/serp.py [hunk 3]
```diff
@@ -48,6 +59,61 @@ class SerpMaxPagesParam(BaseModel):
...
+        description=(
+            "If specified, result URLs are followed to extract the specified "
```
Comment (kmike) [inline zyte_spider_templates/spiders/serp.py:109]: I wonder if we should make the description Google-specific. For example, there are organic and sponsored results; if I get it right, that's organic results which are followed, not sponsored - we should be explicit about it (unless we consider it a bug that sponsored results are not followed - which I'm not sure is a bug).

Comment (kmike) [reply zyte_spider_templates/spiders/serp.py:109]: Something like "Follow the links in the organic search results, and extract the selected data type from the target pages"?

File: zyte_spider_templates/spiders/serp.py [hunk 4]
```diff
@@ -58,6 +124,7 @@ class GoogleDomainParam(BaseModel):
 
 class GoogleSearchSpiderParams(
     MaxRequestsParam,
+    SerpItemTypeParam,
```
Comment (kmike) [inline zyte_spider_templates/spiders/serp.py:127]: Should we provide extract_from option, which applies only for the items extracted from the linked websites, not to search results extraction itself?

Comment (kmike) [reply zyte_spider_templates/spiders/serp.py:127]: Doesn't need to be in the same PR though.
We also should add custom attributes extraction, probably also separately.

**Comment by kmike** [comment]
The implementation looks good to me.

File: zyte_spider_templates/spiders/serp.py (hunk 3 above)
Comment (Gallaecio) [reply zyte_spider_templates/spiders/serp.py]: 953091b?

**Review by kmike — APPROVED** [review]

**Pull request merged**